### PR TITLE
Add Java ACA and AppService templates

### LINKED ISSUE #415: [Feature] Add Java ToDo API
Add a Java version of the ToDo API

### PR DESCRIPTION
**Pull request by weikanglim**
This PR adds SimpleTodo templates for Java AppService and Container Apps. Fixes #415. Note: Java support was added in `azd` with #773. 

Manually tested:
- [x] `azd up, monitor, down` commands
- [x] playwright tests
- [x] VSCode Debugging experience 
- [x] Remote Containers experience

Thanks @jdubois and @brunoborges for the initial SimpleTodo API implementation.

### CODE REVIEW

File: templates/todo/common/infra/bicep/app/api-container-app.bicep [hunk 1]
```diff
@@ -18,6 +18,8 @@ module api '../../../../../common/infra/bicep/core/host/container-app.bicep' = {
     location: location
     containerAppsEnvironmentName: containerAppsEnvironmentName
     containerRegistryName: containerRegistryName
+    containerCpuCoreCount: '1.0'
+    containerMemory: '2.0Gi'
```
Comment (weikanglim) [inline templates/todo/common/infra/bicep/app/api-container-app.bicep:22]: This bumps up the default cores for ACA from 0.5 to 1.0, and memory usage from 1GB to 2GB, which makes it more inline with the AppService plan we use.

File: templates/todo/projects/java-mongo/README.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
+# ToDo Application with a Java API and Azure Cosmos DB API for MongoDB on Azure App Service
+
+[![Open in Remote - Containers](https://img.shields.io/static/v1?label=Remote%20-%20Containers&message=Open&color=blue&logo=visualstudiocode)](https://vscode.dev/redirect?url=vscode://ms-vscode-remote.remote-containers/cloneInVolume?url=https://github.com/azure-samples/todo-nodejs-mongo)
```
Comment (jongio) [inline templates/todo/projects/java-mongo/README.md:3]:
```suggestion
[![Open in Remote - Containers](https://img.shields.io/static/v1?label=Remote%20-%20Containers&message=Open&color=blue&logo=visualstudiocode)](https://vscode.dev/redirect?url=vscode://ms-vscode-remote.remote-containers/cloneInVolume?url=https://github.com/azure-samples/todo-java-mongo)
```

**Review by wbreza — APPROVED** [review]
Looks great overall.  I would recommend someone with more Java experience review the Java specific bits.

**Comment by weikanglim** [comment]
@jdubois, @brunoborges, @roryp or @bbenz if any one of you can help sanity check the Java portion of changes to make sure nothing is amiss, that'd be appreciated.

I'm still waiting to get a `daily` azd build published so that you can run without having to compile a dev build locally. In the meantime, it requires following [contributing.md](https://github.com/Azure/azure-dev/blob/main/cli/azd/CONTRIBUTING.md) to build an `azd` locally..

**Comment by jdubois** [comment]
@weikanglim did you copy/paste the Java application I coded about 6 months ago? If that's the case, it should be all good. Maybe we can update it to the latest Spring Boot 2.x release, that shouldn't change anything (and remove a few CVE advisories).

**Comment by weikanglim** [comment]
@jdubois That is correct. All the Java code was lifted from your existing implementation. Only slight modifications based on app properties, keyvault and appinsights integration.

**Comment by jdubois** [comment]
@weikanglim then that should be all good. I'll have a look to update it to the latest Spring Boot version, but that's not urgent, it's already very recent.

**Comment by roryp** [comment]
I will test today and let you know

regards

Rory

**Comment by roryp** [comment]
It seems the docker device runs out of space for the pr/551

I suggest creating the docker mount with double the storage space.

#4 extracting
sha256:466cf5840e59205c075ee1121d8cfb41cac9c8d64d1fade97d9fa35f34460169 3.9s
#4 ERROR: failed to register layer: Error processing tar file(exit status
1): write /usr/lib/jvm/msopenjdk-17/lib/modules: no space left on device
------
 > [build 1/9] FROM
***@***.***:be60f04045428da1c8604897aaf96d7a178e4f44d07cc2981e6c49dc0a3b24b5
:
------
failed to register layer: Error processing tar file(exit status 1): write
/usr/lib/jvm/msopenjdk-17/lib/modules: no space left on device
: exit status 1
Rorys-MacBook-Pro-3:todo-java-mongo-aca rorypreddy$

**Comment by roryp** [comment]
Full failure log for pr/551 for container apps template:

Rorys-MacBook-Pro-3:todo-java-mongo-aca rorypreddy$ azd up
? Please enter a new environment name: todocosmosrpza
? Please select an Azure location to use: 41. (US) East US 2 (eastus2)
? Please select an Azure Subscription to use: 23. ca-ropreddy-demo-test
(296f38ef-7319-4a2b-ab48-2f38cb21966e)

Infrastructure provisioning plan completed successfully
Provisioning Azure resources can take some time.

You can view detailed progress in the Azure Portal:
https://portal.azure.com/#blade/HubsExtension/DeploymentDetailsBlade/overview/id/%2Fsubscriptions%2F296f38ef-7319-4a2b-ab48-2f38cb21966e%2Fproviders%2FMicrosoft.Resources%2Fdeployments%2Ftodocosmosrpza

Created Resource group: rg-todocosmosrpza
Created Key vault: kv-ozejguycfkida
Created Log Analytics workspace: log-ozejguycfkida
Created Application Insights: appi-ozejguycfkida
Created Portal dashboard: dash-ozejguycfkida
Created Container Apps Environment: cae-ozejguycfkida
Created Container App: ca-api-ozejguycfkida
Created Azure Cosmos DB: cosmos-ozejguycfkida
Created Container App: ca-web-ozejguycfkida

Azure resource provisioning completed successfully

Deploying service api...
Error: deploying service: packaging service api: building container: api at
.: building image: exit code: 1, stdout: , stderr: #1 [internal] load build
definition from Dockerfile
#1 sha256:fea6948161f7e274462dc7e31ba847b9d8313b1bb5da0f8bfff5969805da2c4c
#1 transferring dockerfile: 777B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 sha256:86779696bd4f4c763a30f89cd74f33ce2a9dce8c7318b574cc918e1807f0460d
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for mcr.microsoft.com/openjdk/jdk:17-mariner
#3 sha256:12da1caa7cb1e091da203d54e325c61d9dc8868901554bda381e714294745d9a
#3 DONE 3.7s

#6 [internal] load build context
#6 sha256:7e1d23fea746ca32373c4dd81c12ef1f6db700441d7bd37e8a50243d2ec39958
#6 transferring context: 132.34kB 0.0s done
#6 DONE 0.0s

#4 [build 1/9] FROM
***@***.***:be60f04045428da1c8604897aaf96d7a178e4f44d07cc2981e6c49dc0a3b24b5
#4 sha256:8c1a155b388143167348573bd9e32802c92d6857d1230f43b95dc2583eb670b9
#4 resolve
***@***.***:be60f04045428da1c8604897aaf96d7a178e4f44d07cc2981e6c49dc0a3b24b5
done
#4 sha256:466cf5840e59205c075ee1121d8cfb41cac9c8d64d1fade97d9fa35f34460169
149.69MB / 149.69MB 112.3s done
#4 sha256:be60f04045428da1c8604897aaf96d7a178e4f44d07cc2981e6c49dc0a3b24b5
742B / 742B done
#4 sha256:1e4ab1988ec292ad33262cfc7922458c62c8ecc5f0b43e8a518bdad9688f9c7c
5.88kB / 5.88kB done
#4 sha256:7f0517a4d0b40accc745957dd7518369b86e064df31d5c75fd94af60b0dd5e4e
28.09MB / 28.09MB 15.2s done
#4 extracting
sha256:7f0517a4d0b40accc745957dd7518369b86e064df31d5c75fd94af60b0dd5e4e
1.1s done
#4 extracting
sha256:466cf5840e59205c075ee1121d8cfb41cac9c8d64d1fade97d9fa35f34460169 3.9s
#4 ERROR: failed to register layer: Error processing tar file(exit status
1): write /usr/lib/jvm/msopenjdk-17/lib/modules: no space left on device
------
 > [build 1/9] FROM
***@***.***:be60f04045428da1c8604897aaf96d7a178e4f44d07cc2981e6c49dc0a3b24b5
:
------
failed to register layer: Error processing tar file(exit status 1): write
/usr/lib/jvm/msopenjdk-17/lib/modules: no space left on device
: exit code: 1, stdout: , stderr: #1 [internal] load build definition from
Dockerfile
#1 sha256:fea6948161f7e274462dc7e31ba847b9d8313b1bb5da0f8bfff5969805da2c4c
#1 transferring dockerfile: 777B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 sha256:86779696bd4f4c763a30f89cd74f33ce2a9dce8c7318b574cc918e1807f0460d
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for mcr.microsoft.com/openjdk/jdk:17-mariner
#3 sha256:12da1caa7cb1e091da203d54e325c61d9dc8868901554bda381e714294745d9a
#3 DONE 3.7s

#6 [internal] load build context
#6 sha256:7e1d23fea746ca32373c4dd81c12ef1f6db700441d7bd37e8a50243d2ec39958
#6 transferring context: 132.34kB 0.0s done
#6 DONE 0.0s

#4 [build 1/9] FROM
***@***.***:be60f04045428da1c8604897aaf96d7a178e4f44d07cc2981e6c49dc0a3b24b5
#4 sha256:8c1a155b388143167348573bd9e32802c92d6857d1230f43b95dc2583eb670b9
#4 resolve
***@***.***:be60f04045428da1c8604897aaf96d7a178e4f44d07cc2981e6c49dc0a3b24b5
done
#4 sha256:466cf5840e59205c075ee1121d8cfb41cac9c8d64d1fade97d9fa35f34460169
149.69MB / 149.69MB 112.3s done
#4 sha256:be60f04045428da1c8604897aaf96d7a178e4f44d07cc2981e6c49dc0a3b24b5
742B / 742B done
#4 sha256:1e4ab1988ec292ad33262cfc7922458c62c8ecc5f0b43e8a518bdad9688f9c7c
5.88kB / 5.88kB done
#4 sha256:7f0517a4d0b40accc745957dd7518369b86e064df31d5c75fd94af60b0dd5e4e
28.09MB / 28.09MB 15.2s done
#4 extracting
sha256:7f0517a4d0b40accc745957dd7518369b86e064df31d5c75fd94af60b0dd5e4e
1.1s done
#4 extracting
sha256:466cf5840e59205c075ee1121d8cfb41cac9c8d64d1fade97d9fa35f34460169 3.9s
#4 ERROR: failed to register layer: Error processing tar file(exit status
1): write /usr/lib/jvm/msopenjdk-17/lib/modules: no space left on device
------
 > [build 1/9] FROM
***@***.***:be60f04045428da1c8604897aaf96d7a178e4f44d07cc2981e6c49dc0a3b24b5
:
------
failed to register layer: Error processing tar file(exit status 1): write
/usr/lib/jvm/msopenjdk-17/lib/modules: no space left on device
: exit status 1

**Comment by roryp** [comment]
Created Container Apps Environment: cae-ozejguycfkida
Error: deploying infrastructure: error deploying infrastructure: failed
deploying: failed running az deployment sub create:

Deployment Error Details:
ConflictError: A vault with the same name already exists in deleted state.
You need to either recover or purge existing key vault. Follow this link
https://go.microsoft.com/fwlink/?linkid=2149745 for more information on
soft delete.next issue: I tried to manually delete the resource group after
the failure and cant recreate the app as the keyvault needs to be manually
deleted:

**Comment by roryp** [comment]
I just tested the github actions via a private repo and it works great! When will you publish the live template @weikanglim ?
image

File: templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/web/TodoListsController.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                .toUri();
+        return ResponseEntity.created(location).body(savedTodoList);
+    }
+
+    @Override
+    public ResponseEntity<Void> deleteItemById(String listId, String itemId) {
+        Optional<TodoItem> todoItem = getTodoItem(listId, itemId);
+        if (todoItem.isPresent()) {
+            todoItemRepository.deleteById(itemId);
+            return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
```
Comment (wangmingliang-ms) [inline templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/web/TodoListsController.java:65]: ResponseEntity.noContent().build();

File: templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/model/TodoState.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  @Override
+  public String toString() {
+    return String.valueOf(value);
+  }
+
+  @JsonCreator
+  public static TodoState fromValue(String value) {
+    for (TodoState b : TodoState.values()) {
+      if (b.value.equals(value)) {
```
Comment (wangmingliang-ms) [inline templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/model/TodoState.java:51]: I would recommend to use `equalsIgnoreCase()`

File: templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/web/TodoListsController.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            todoItemRepository.save(todoItem);
+        }
+        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+    }
+
+    @Override
+    public ResponseEntity<TodoList> updateListById(String listId, TodoList todoList) {
+        return todoListRepository
+                .findById(listId)
+                .map(t -> ResponseEntity.ok(todoListRepository.save(t)))
```
[the remark this answers is not in the document]
Comment (wangmingliang-ms) [reply templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/web/TodoListsController.java:153]: we should also make sure the `listId` is set into `todoList` before saving.

File: templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/web/TodoListsController.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    @Override
+    public ResponseEntity<TodoItem> updateItemById(String listId, String itemId, TodoItem todoItem) {
+        return getTodoItem(listId, itemId).map(t -> {
+            todoItemRepository.save(todoItem);
+            return ResponseEntity.ok(todoItem);
+        }).orElseGet(() -> ResponseEntity.notFound().build());
+    }
+
+    @Override
+    public ResponseEntity<Void> updateItemsStateByListId(String listId, TodoState state, List<String> requestBody) {
```
Comment (wangmingliang-ms) [inline templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/web/TodoListsController.java:141]: remove `requestBody` if not used.

**Comment by jdubois** [comment]
@wangmingliang-ms I agree with your PR review but be careful: many of the issues come because we inherit com.microsoft.azure.simpletodo.api.ListsApi which clearly isn't perfect, but is automatically generated using https://openapi-generator.tech (and to be more specific: their Spring code generator).

As a result, unfortunately some of the methods cannot be changed, as they are inherited from this base class: sometimes that's because our Swagger contract isn't good (I found 3 errors in the contract that way), sometimes it's because OpenAPI doesn't generate perfect code (but we can contribute back to the project and fix it).

**Comment by jongio** [comment]
@wangmingliang-ms - Would you like to do a PR for these changes?  Thanks!  Jon

**Comment by wangmingliang-ms** [comment]
> @wangmingliang-ms - Would you like to do a PR for these changes? Thanks! Jon

sure, I am happy to do that.

**Comment by jongio** [comment]
Thanks you!

**Comment by wangmingliang-ms** [comment]
> @wangmingliang-ms - Would you like to do a PR for these changes? Thanks! Jon

Hi, @jongio  here is the PR: https://github.com/Azure/azure-dev/pull/1122, it also fixes some other issues in addition to the above.

**Pull request merged**